include/mbgl/gfx/drawable_builder.hpp

```diff
@@ -60,6 +60,11 @@ class DrawableBuilder {
     const Color& getColor() const;
     void setColor(const Color& value);
 
+    enum class ColorMode { PerDrawable, PerVertex };
+    /// Set how the color value is used
+    /// This should not be changed while a build is in progress
+    void setColorMode(ColorMode mode) { colorMode = mode; }
```

Comment:
Does it make sense to have the concept of color in the builder like this? There are so many paint properties and they vary per layer, it feels like it would make more sense to have stuff like this in the respective layer code using the builder and decouple the builder from properties.

Take a look [here](https://maplibre.org/maplibre-style-spec/layers/#layout-property), note the difference between "layout" and "paint" properties. Properties with the palette icon are paint properties.

Reply:
I don't know, it's a placeholder that will be replaced, but I'm not sure by what.
Should the drawables carry along the evaluated paint properties?  That stuff all seems to be templated, so isn't easy to pass around generically.
I don't think we want to create new ways of representing all the properties that can be dynamic, though.

Reply:
I guess there are other things that can be per-vertex vs per-drawable, so that could be more generic.
As for changing these only the fly, I like how you're putting that logic into the Tweaker.  But the Drawable should probably have a way of setting new values and then updating when dirty.

Reply:
WhirlyGlobe/Maply translates the style paint properties into `Expression` objects which are further translated into shader inputs to be evaluated in the shaders on Metal.  But it only supports a subset of the available properties.
For anything we want to be evaluated that dynamically, obviously we can't avoid at least one additional representation.
We probably also want to be careful to avoid re-evaluating any layer properties for each drawable.

Reply:
In Metal we're going to do most if not all of this calculation in the shaders so it makes sense to turn it into a generic representation which can then either be run in a Tweaker or translated to data which is passed to a shader.

Reply:
Currently, drawables have tweakers that get run, but I think we'll have tweakers that apply their results to a collection of drawables, so it might make sense to invert that relation, or add a new mechanism for it.

Reply:
I'd say do it naively to start and then see how it looks.
Rather than putting effort into a CPU side solution it might just make more sense to put it into the shaders, even for OpenGL.

Reply:
Looking into this more, it doesn't really seem practical to pass the paint parameters around through generic drawables/tweakers, but I don't think we need to.  The layer update method that generates new tile drawables when necessary is also the right place to update those drawables with new results for evaluated properties.

Reply:
That makes a lot of sense for this phase.  I think we'll want to revisit that approach in the next phase when we get to the multi-threading exploration.  If we have time to move some of this logic into the Metal shaders that will help some too.

Reply:
I do want to emphasize you shouldn't worry about multi-threading quite yet.
I can see we're going to do a (mostly) single threaded implementation for Metal at first.  Once that's working, then we can come back to some of these issues and figure out how to make the toolkit truly multi-threaded.

Reply:
Agreed.  If we can make the property evaluation async I think we'd also do the translation of the evaluated properties into attribute values at the same time, and then apply them via change requests, so we just need to be able to pass attribute values/arrays through generically.

Reply:
That makes sense.